File: merge-tonic-lib/tonic/repo_profile.py
"""Read/write `.tonic/repo.json` (tonic-repo-profile v1) with idempotent merge semantics."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, TypedDict

TONIC_REPO_PROFILE_SCHEMA = "tonic-repo-profile"
TONIC_REPO_PROFILE_VERSION = "1"
DEFAULT_REPO_PROFILE_REL = Path(".tonic") / "repo.json"


class TonicRepoProfileV1(TypedDict, total=False):
    schema: str
    version: str
    remote: str
    canonical_ref: str
    left_ref: str
    right_ref: str
    intent_pair: str
    intent_profile: str
    hydrate_out_dir: str
    git_remote_url: str
    hub_repo_id: str
    hub_default_dag: bool
    replay_trace_source: str
    hub_index_path: str
    compare_mode: str


def default_repo_profile() -> TonicRepoProfileV1:
    return {
        "schema": TONIC_REPO_PROFILE_SCHEMA,
        "version": TONIC_REPO_PROFILE_VERSION,
        "remote": "origin",
        "canonical_ref": "",
        "left_ref": "",
        "right_ref": "",
        "intent_pair": "",
        "intent_profile": "",
        "hydrate_out_dir": "",
        "git_remote_url": "",
        "hub_repo_id": "",
        "hub_default_dag": False,
        "replay_trace_source": "auto",
        "hub_index_path": "",
        "compare_mode": "snapshot",
    }


def parse_repo_profile_json(raw: str) -> TonicRepoProfileV1:
    try:
        data: Any = json.loads(raw)
    except json.JSONDecodeError as e:
        raise ValueError("repo profile: invalid JSON") from e
    if not isinstance(data, dict):
        raise ValueError("repo profile: root must be object")
    if data.get("schema") != TONIC_REPO_PROFILE_SCHEMA:
        raise ValueError(f"repo profile: schema must be {TONIC_REPO_PROFILE_SCHEMA}")
    if data.get("version") != TONIC_REPO_PROFILE_VERSION:
        raise ValueError(f"repo profile: version must be {TONIC_REPO_PROFILE_VERSION}")
    remote = data.get("remote", "")
    if not isinstance(remote, str) or not remote.strip():
        raise ValueError("repo profile: remote must be non-empty string")
    cm = data.get("compare_mode", "snapshot")
    if cm not in ("snapshot", "weave", "", None):
        raise ValueError('repo profile: compare_mode must be "snapshot" or "weave"')
    base = default_repo_profile()
    extra = (
        "git_remote_url",
        "hub_repo_id",
        "hub_default_dag",
        "replay_trace_source",
        "hub_index_path",
    )
    for k, v in data.items():
        if k in base or k in extra:
            base[k] = v  # type: ignore[literal-required]
    base["remote"] = str(remote).strip()
    return base


def merge_repo_profiles(base: TonicRepoProfileV1, patch: dict[str, Any]) -> TonicRepoProfileV1:
    out: dict[str, Any] = {**default_repo_profile(), **base}
    for k, v in patch.items():
        if v is None:
            continue
        if isinstance(v, str):
            if v.strip():
                out[k] = v.strip()
            continue
        if k == "compare_mode" and v in ("snapshot", "weave"):
            out[k] = v
            continue
        if k == "hub_default_dag" and isinstance(v, bool):
            out[k] = v
            continue
        if k == "replay_trace_source" and v in ("local", "hub", "auto"):
            out[k] = v
            continue
        if k == "hub_index_path" and isinstance(v, str) and v.strip():
            out[k] = v.strip()
    out["schema"] = TONIC_REPO_PROFILE_SCHEMA
    out["version"] = TONIC_REPO_PROFILE_VERSION
    return out  # type: ignore[return-value]


def repo_profile_path(repo_root: str | Path, rel: Path | None = None) -> Path:
    r = Path(repo_root).resolve()
    return r / (rel or DEFAULT_REPO_PROFILE_REL)


def read_repo_profile(repo_root: str | Path) -> TonicRepoProfileV1 | None:
    p = repo_profile_path(repo_root)
    if not p.is_file():
        return None
    return parse_repo_profile_json(p.read_text(encoding="utf-8"))


def write_repo_profile(repo_root: str | Path, profile: TonicRepoProfileV1, rel: Path | None = None) -> None:
    p = repo_profile_path(repo_root, rel)
    p.parent.mkdir(parents=True, exist_ok=True)
    body = json.dumps(profile, indent=2, sort_keys=True) + "\n"
    p.write_text(body, encoding="utf-8")


def refs_to_fetch_from_profile(p: TonicRepoProfileV1) -> list[str]:
    s: set[str] = set()
    for r in (p.get("canonical_ref"), p.get("left_ref"), p.get("right_ref")):
        t = (r or "").strip()
        if t:
            s.add(t)
    return sorted(s)

File: packages/tonic-core/src/repoProfile/repoProfile.ts
import * as fs from "node:fs";
import * as path from "node:path";

export const TONIC_REPO_PROFILE_SCHEMA = "tonic-repo-profile" as const;
export const TONIC_REPO_PROFILE_VERSION = "1" as const;
export const DEFAULT_REPO_PROFILE_REL = path.join(".tonic", "repo.json");

export type TonicRepoProfileV1 = {
  schema: typeof TONIC_REPO_PROFILE_SCHEMA;
  version: typeof TONIC_REPO_PROFILE_VERSION;
  remote: string;
  canonical_ref?: string;
  left_ref?: string;
  right_ref?: string;
  intent_pair?: string;
  intent_profile?: string;
  hydrate_out_dir?: string;
  git_remote_url?: string;
  hub_repo_id?: string;
  compare_mode?: "snapshot" | "weave";
};

export function defaultRepoProfile(): TonicRepoProfileV1 {
  return {
    schema: TONIC_REPO_PROFILE_SCHEMA,
    version: TONIC_REPO_PROFILE_VERSION,
    remote: "origin",
    canonical_ref: "",
    left_ref: "",
    right_ref: "",
    intent_pair: "",
    intent_profile: "",
    hydrate_out_dir: "",
    git_remote_url: "",
    hub_repo_id: "",
    compare_mode: "snapshot",
  };
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** Parse and validate minimal repo profile JSON. */
export function parseRepoProfileJson(raw: string): TonicRepoProfileV1 {
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch {
    throw new Error("repo profile: invalid JSON");
  }
  if (!isRecord(data)) {
    throw new Error("repo profile: root must be object");
  }
  if (data.schema !== TONIC_REPO_PROFILE_SCHEMA) {
    throw new Error(`repo profile: schema must be ${TONIC_REPO_PROFILE_SCHEMA}`);
  }
  if (data.version !== TONIC_REPO_PROFILE_VERSION) {
    throw new Error(`repo profile: version must be ${TONIC_REPO_PROFILE_VERSION}`);
  }
  if (typeof data.remote !== "string" || !data.remote.trim()) {
    throw new Error("repo profile: remote must be non-empty string");
  }
  const cm = data.compare_mode;
  if (cm !== undefined && cm !== "snapshot" && cm !== "weave") {
    throw new Error('repo profile: compare_mode must be "snapshot" or "weave"');
  }
  const base = defaultRepoProfile();
  for (const k of Object.keys(data)) {
    if (k in base || k === "git_remote_url" || k === "hub_repo_id") {
      (base as Record<string, unknown>)[k] = data[k];
    }
  }
  base.remote = String(data.remote).trim();
  return base as TonicRepoProfileV1;
}

/** Deep merge: non-empty string values in `patch` override `base`. */
export function mergeRepoProfiles(base: TonicRepoProfileV1, patch: Partial<TonicRepoProfileV1>): TonicRepoProfileV1 {
  const out: TonicRepoProfileV1 = { ...defaultRepoProfile(), ...base };
  for (const [k, v] of Object.entries(patch)) {
    if (v === undefined) {
      continue;
    }
    if (typeof v === "string") {
      if (v.trim() !== "") {
        (out as Record<string, unknown>)[k] = v.trim();
      }
      continue;
    }
    if (k === "compare_mode" && (v === "snapshot" || v === "weave")) {
      out.compare_mode = v;
    }
  }
  out.schema = TONIC_REPO_PROFILE_SCHEMA;
  out.version = TONIC_REPO_PROFILE_VERSION;
  return out;
}

export function repoProfilePath(repoRoot: string, rel = DEFAULT_REPO_PROFILE_REL): string {
  return path.join(repoRoot, rel);
}

export function readRepoProfile(repoRoot: string): TonicRepoProfileV1 | null {
  const p = repoProfilePath(repoRoot);
  if (!fs.existsSync(p)) {
    return null;
  }
  return parseRepoProfileJson(fs.readFileSync(p, "utf8"));
}

export function writeRepoProfile(repoRoot: string, profile: TonicRepoProfileV1, rel = DEFAULT_REPO_PROFILE_REL): void {
  const p = repoProfilePath(repoRoot, rel);
  fs.mkdirSync(path.dirname(p), { recursive: true });
  const body = JSON.stringify(profile, null, 2) + "\n";
  fs.writeFileSync(p, body, "utf8");
}

/** All ref-like strings from profile used for fetch (deduped, non-empty). */
export function refsToFetchFromProfile(p: TonicRepoProfileV1): string[] {
  const s = new Set<string>();
  for (const r of [p.canonical_ref, p.left_ref, p.right_ref]) {
    const t = (r ?? "").trim();
    if (t) {
      s.add(t);
    }
  }
  return [...s];
}
